File: leetcode/src/main/java/com/giwankim/leetcode/Trie.java
package com.giwankim.leetcode;

public class Trie {
  private final TrieNode root;

  public Trie() {
    root = new TrieNode();
  }

  public void insert(String word) {
    TrieNode cur = root;
    for (char c : word.toLowerCase().toCharArray()) {
      if (cur.children[c - 'a'] == null) {
        cur.children[c - 'a'] = new TrieNode();
      }
      cur = cur.children[c - 'a'];
    }
    cur.isWord = true;
  }

  public boolean search(String word) {
    TrieNode cur = root;
    for (char c : word.toLowerCase().toCharArray()) {
      if (cur.children[c - 'a'] == null) {
        return false;
      }
      cur = cur.children[c - 'a'];
    }
    return cur.isWord;
  }

  public boolean startsWith(String prefix) {
    TrieNode cur = root;
    for (char c : prefix.toLowerCase().toCharArray()) {
      if (cur.children[c - 'a'] == null) {
        return false;
      }
      cur = cur.children[c - 'a'];
    }
    return true;
  }

  public static class TrieNode {
    private boolean isWord;
    private final TrieNode[] children;

    public TrieNode() {
      this.isWord = false;
      this.children = new TrieNode[26];
    }
  }
}

File: leetcode-kt/src/main/kotlin/com/giwankim/leetcode/Trie.kt
package com.giwankim.leetcode

class TrieNode {
    var isWord: Boolean = false
    val children: Array<TrieNode?> = arrayOfNulls(26)

    fun contains(c: Char): Boolean = children[c - 'a'] != null

    fun child(c: Char): TrieNode? = children[c - 'a']
}

class Trie {
    val root: TrieNode = TrieNode()

    fun insert(word: String) {
        var cur = root
        for (c in word.lowercase()) {
            if (!cur.contains(c)) {
                cur.children[c - 'a'] = TrieNode()
            }
            cur = cur.child(c)!!
        }
        cur.isWord = true
    }

    fun search(word: String): Boolean {
        var cur: TrieNode = root
        for (c in word.lowercase()) {
            if (!cur.contains(c)) {
                return false
            }
            cur = cur.child(c)!!
        }
        return cur.isWord
    }

    fun startsWith(prefix: String): Boolean {
        var cur: TrieNode = root
        for (c in prefix.lowercase()) {
            if (!cur.contains(c)) {
                return false
            }
            cur = cur.child(c)!!
        }
        return true
    }
}
